### PR TITLE
Improving Recorded Future Connector (#issue/1418)

### PR DESCRIPTION
**Pull request by helene-nguyen**
### Proposed changes

* Handle multiple topics
* Add a parameter "risk_as_score" to use risk score (RF) as score for indicators (Hash, IP, domain)
* Add a parameter "risk_threshold" under which related indicators are not taken into account
* Mapping more related entities 
  * Product -> Software, 
  * Mitre Attack Pattern -> Attack pattern, 
  * City, Country, ProvinceOrState -> City, Country, Administrative-Area
  * Industry -> Sector
  * Operation -> Campaign
* Mapping Sigma rules
* Add relationships between entities contained in a report

image

### Related issues

* https://github.com/OpenCTI-Platform/connectors/issues/1418

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### CODE REVIEW

**Review by marieflorescontact — COMMENTED** [review]
locally tested, some questions:

- is it normal to have indicators with `labels":["informational","validated intelligence event"]` with config topic option set to `topic: 'UrMRnT'`? should'nt it be  `labels":["informational"]` ?

- For `risk_threshold` config set to `risk_threshold: 70` we have some URL `stix` indicators with `score : 50/100` (discuss with @helene-nguyen)

File: external-import/recordedfuture-notes/src/rflib/rf_notes_to_stix2.py [hunk 1]
```diff
@@ -313,9 +319,68 @@ def create_stix_objects(self):
...
+        self.software_object = stix2.Software(
+            name=self.name,
+        )
+
+
+class Location(RFStixEntity):
+    rf_type_to_stix = {
+        "Country": "Country",
+        "City": "City",
+        "ProvinceOrState": "Administrative-Area",
```
Comment (marieflorescontact) [inline external-import/recordedfuture-notes/src/rflib/rf_notes_to_stix2.py:343]: Why not `Region` in `Location` class ? (I have seen the Entities Mapping Relationships schema, where it is specified `Country/City/Adm Area` but i am curious)

File: external-import/recordedfuture-notes/src/rflib/rf_notes_to_stix2.py [hunk 2]
```diff
@@ -420,21 +505,131 @@ def from_json(self, note):
...
+                {"entity": "attack-pattern", "relation": "uses"},
+                {"entity": "location", "relation": "targets"},
+                {"entity": "identity", "relation": "targets"},
+            ],
+        },
+        {
+            "from": "indicator",
+            "to": [
+                {"entity": "malware", "relation": "indicates"},
+                {"entity": "threat-actor", "relation": "indicates"},
```
Comment (marieflorescontact) [inline external-import/recordedfuture-notes/src/rflib/rf_notes_to_stix2.py:580]: Did you miss `{"entity": "intrusion-set", "relation": "indicates"}` ? (based on Entities mapping relationships schema)

**Comment by helene-nguyen** [comment]
> locally tested, some functionnal questions, I'll look into the code tomorrow morning:
> 
> * is it normal to have indicators with `labels":["informational","validated intelligence event"]` with config topic option set to `topic: 'UrMRnT'`? shouldn't it be  `labels":["informational"]` ?
> * For `risk_threshold` config set to `risk_threshold: 70` we have some URL `stix` indicators with `score : 50/100` (discuss with @helene-nguyen)

For the first part, yes it is normal, the indicator will be imported with its labels and if  `informational` exists, all labels will be imported

For the 2nd part, follow-up our discussion, URL is not handled yet but we are on the way to do it :)

File: external-import/recordedfuture-notes/src/rflib/rf_notes_to_stix2.py (hunk 2 above)
Comment (helene-nguyen) [reply external-import/recordedfuture-notes/src/rflib/rf_notes_to_stix2.py:580]: Yes, thank you! 
Done :)

File: external-import/recordedfuture-notes/src/rflib/rf_notes_to_stix2.py (hunk 1 above)
Comment (helene-nguyen) [reply external-import/recordedfuture-notes/src/rflib/rf_notes_to_stix2.py:343]: There are some Province or State which is not a region, it is a choice to map with Administrative Area but need to be discussed with @Jipegien

**Review by marieflorescontact — COMMENTED** [review]
as discussed with @helene-nguyen 

error when starting connector:

`{"timestamp": "2023-11-06T08:30:36.537119Z", "level": "ERROR", "name": "pycti.connector", "message": "401 Client Error: Unauthorized for url: https://api.recordedfuture.com/detection-rule/search"}`

**Comment by yassine-ouaamou** [comment]
Update: Concerning the error that Marie encountered, a ticket has been created for the API team in Recorded Future.

**Review by marieflorescontact — APPROVED** [review]
locally tested ok 👌

**Pull request merged**